### ISSUE TITLE
is volatile keyword needed in even-odd wait notify example

### ISSUE THREAD

**Issue by p1729** [eugenp/tutorials#7806]
I was going through your tutorial on https://www.baeldung.com/java-even-odd-numbers-with-2-threads section 5.1, is the volatile keyword necessary ? as the flag has been checked in the same instance
https://github.com/eugenp/tutorials/blob/9f2f00e1b624b7eab3c0ce1778404aa217ff6005/core-java-modules/core-java-concurrency-advanced/src/main/java/com/baeldung/concurrent/evenandodd/PrintEvenOddWaitNotify.java#L40

**Comment by alimate** [comment]
> is the volatile keyword necessary ? as the flag has been checked in the same instance

Based on this argument, the answer is **No**. The same `Printer` instance is being shared between two threads. Suppose those two threads are running on two physical cores and after a while they both cache the `isOdd` value on their dedicated caches:
 - Thread 2 updates the `isOdd` value.
 - After that, Thread 1 reads the `isOdd`. **There is no guarantee that this value is the recent write.**

Happens Before
-----------------
In order to make sure that any write to the `isOdd` variable _happens before_ any read after that, we can make the variable a _volatile_ one.

Piggybacking on Synchronization
---------------
Both methods are `synchronized`:
```
class Printer {
    private volatile boolean isOdd;
 
    synchronized void printEven(int number) {
        while (!isOdd) {
            // omitted
        }
        // omitted
        isOdd = false;
        notify();
    }
 
    synchronized void printOdd(int number) {
        while (isOdd) {
            // omitted
        }
        // omitted
        isOdd = true;
        notify();
    }
}
```
In addition to `volatile`, any unlock on a monitor lock _happens before_ all subsequent locks on the same lock. So, all writes to `isOdd` after exiting from the `synchronized` are visible to all subsequent reads when entering a new `synchronized`. 

**Put simply, the `isOdd` does not need to be a volatile one.**
